2_idioms/src/main.rs

```diff
@@ -1,3 +1,452 @@
+                }
+                for item in items {
+                    product.push(item)
+                }
+                Ok(())
+            }
+            None => {
+                if self.product_storage.len() == self.stock_capacity.0 as usize {
+                    return Err("There's no empty positions in stock");
+                }
+                if items.len() > self.in_stock_capacity.0 as usize {
+                    return Err("Cannot add that much");
+                }
+                let mut new_vec: Vec<Box<dyn Product>> = Vec::new();
+                for item in items {
+                    new_vec.push(item);
+                }
+                self.product_storage.insert(TypeId::of::<T>(), new_vec);
+                return Ok(());
+            }
+        }
+    }
+
+    pub fn get_price_and_count<T: Product>(&self) -> Option<(u64, usize)> {
+        let tid = TypeId::of::<T>();
+        match self.product_storage.get(&tid) {
+            None => None,
+            Some(product) => Some((product[0].value(), product.len())),
+        }
+    }
+
+    pub fn purchase<T: Product, TAmmount: Into<product::Amount>>(
+        &mut self,
+        amount: TAmmount,
+    ) -> Vec<Box<dyn Product>> {
+        let tid = TypeId::of::<T>();
+        let count = amount.into().0 as usize;
+        let items = self.product_storage.get_mut(&tid).unwrap();
+
+        let drain = items.drain(0..count);
+        let mut vec = Vec::new();
+        for item in drain {
+            vec.push(item);
+        }
+        if items.len() == 0 {
+            self.product_storage.remove(&tid);
+        }
+        vec
+    }
+}
+struct VendingMachine {
+    stock: Stock,
+    rest: Money,
+}
+#[derive(Clone, Copy)]
+struct RowCount(u8);
+#[derive(Clone, Copy)]
+struct InRowCount(u8);
+impl VendingMachine {
+    pub fn display_stock(&self) {
+        self.stock.display_stock()
+    }
+    pub fn new(stock_capacity: RowCount, in_stock_capacity: InRowCount) -> Self {
+        VendingMachine {
+            stock: Stock::new(stock_capacity, in_stock_capacity),
+            rest: Money::new(),
+        }
+    }
+
+    pub fn insert<T: Product, TCollection: IntoIterator<Item = T>>(
+        &mut self,
+        items: TCollection,
+    ) -> Result<(), &'static str> {
+        let mut collection = Vec::new();
+        for item in items {
+            collection.push(Box::new(item))
+        }
+        self.stock.insert(collection)
+    }
+
+    pub fn add_money(&mut self, money: money::Money) {
+        self.rest.merge(money);
+    }
+
+    pub fn purchase<T: Product, TAmmount: Into<product::Amount>>(
+        &mut self,
+        money: money::Money,
+        amount: TAmmount,
+    ) -> Result<(Vec<Box<dyn Product>>, Money), (Money, &'static str)> {
+        let items_amount = amount.into();
+        match self.stock.get_price_and_count::<T>() {
+            None => return Err((money, "there's no item of this type")),
```

Comment:
@ArsileLuci use `.map_err()` and `?` operator instead of manual matching.

Reply:
@tyranron do you mean `.map_or()`? Because `.get_price_and_count()` rerutns `Option<(u64, usize)>` and `Option<T>` doesn't have `.map_err()` function.

Reply:
@ArsileLuci yes